### ISSUE TITLE
Provide an ability to add custom actions to the test controller

### ISSUE THREAD

**Issue by AlexanderMoskovkin** [DevExpress/testcafe#1535]
Based on discussions: 
https://testcafe-discuss.devexpress.com/t/use-custom-actions-on-test-controller/228
https://testcafe-discuss.devexpress.com/t/how-to-chain-pagemodel-methods/313

### Are you requesting a feature or reporting a bug?

Feature

### What is the expected behavior?

Provide an ability to add custom actions to the test controller or chain methods that contain a test controller chain

**Comment by stefanfisk** [comment]
Might a simpler alternative be to add a `apply()` action that takes a function as its argument, and then calls that function with the controller as its argument? Then there would be no risk of polluting the controller's API, and we could all factor out whatever we'd like into separate functions, at least provided we only want to do sync calls.

**Comment by stalniy** [comment]
It's already possible with function composition :) You can use smth this:

```js
test('can login', pipe(
  attemptsTo(login('email@example.com', '1234567')),
  expectsTo(see(Dashboard.container))
))
```

**Comment by Hultner** [comment]
Just thinking out loud here but it looks like a custom action could be implemented as a subset of the current `Role` behaviour where a chain of actions are applied when using `.useRole(roleX)`. 

Let's imagine we have a similar `.useAction(customAction)` function and the custom actions would look something like this
```js
const customAction = Action( async(t) => { 
    await t
        .expect(something).ok(…)
        .click(Selector(x))
        .typeText('xyz')
        …
}
// Or just
async function(t){ 
    await t …
}
```

This would provide consistency with current API and wouldn't clutter the t object with our custom methods while still providing good facilities to execute them in the chain.

I could so how in a more mature project an internal library of actions could be procured so the individual tests are much higher level.

**Comment by orihomie** [comment]
Any updates on this?

**Comment by AlexSkorkin** [comment]
Once we get any results, we will post them here.

**Comment by qrush** [comment]
Bumping this, we've been looking for a way to do this with our testcafe suite @wistia, and we'd love to see this merged! If there's anything we can do to help here, let us know!

**Comment by Benyon** [comment]
Is there any update on this since feature?

**Comment by arubtsov** [comment]
@Benyon, We have no updates at the moment.

**Comment by AndreyBelym** [comment]
Another point for discussing this feature:
Using Promise subclasses as return values for re-evaluating functions is bad in modern JS. Since if you return an instance of a Promise subclass from `async` function, it will be awaited automatically inside the engine and wrapper in a regular Promise.
```js
> B = class extends Promise {}
[class B extends Promise]
> x = (async () => new B(resolve => resolve()))()
Promise { <pending> }
> x instanceof B
false
```

This is already especially bad for people using TestCafe and the [promise-function-async](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/promise-function-async.md) TypeScript-ESlint rule. Since ClientFunctions returns extended Promises and the rule can auto-fix your code, using it can turn reexecutable promises into regular promises, practically disabling Smart Assertions and in some cases even completely breaking your test scenario if it is written with chained actions.

**Comment by magic-m-johnson** [comment]
would really like that feature! :)

**Comment by sergiubologa** [comment]
This feature would be gold 🥇 My tests would look awesome. ✨

**Comment by Matenko** [comment]
Definitely would be nice as a few of our tests can be quite long and chaining within single models would make things a lot cleaner.

**Comment by NedWilbur** [comment]
Having a feature such as this would be great when using POM. Without this, I have to break the test chain often and produces messy code. A simple function that just executes whatever is passed to it would make it much easier.

Current:
``` ts
    await t
        .click(Page.Inventory.El.ItemAddRemoveButton)
        .expect(Page.Inventory.El.ItemAddRemoveButton.innerText).eql('REMOVE')
        .expect(Page.Navbar.El.ShoppingCartBadge.innerText).eql('1')
        .click(Page.Navbar.El.ShoppingCartButton);
    await Page.Cart.ItemExist(item);
    await t.click(Page.Cart.El.CheckoutButton)
    await Page.Checkout.EnterInfo(person);
    await Page.Checkout.ItemExist(item);
    await t.click(Page.Checkout.El.FinishButton)
        .expect(Page.Checkout.El.CompleteHeader.innerText).eql('THANK YOU FOR YOUR ORDER');

**Comment by DanKaplanSES** [comment]
I'm not sure why I seem to be the only one on the Internet who ran into this problem, but I kept getting `ReferenceError: Selector is not defined` in my custom action. The solution is to put this at the top of your `.testcaferc` file:

```js
const { Selector } = require("testcafe");
```

Maybe this'll save someone time in the future.

**Comment by miherlosev** [comment]
[https://github.com/DevExpress/testcafe/issues/1535](https://github.com/DevExpress/testcafe/issues/1535)

Hi @DanKaplanSES,

This is how Node.js works: you need to import the `Selector` class before using it.

**Comment by DanKaplanSES** [comment]
> This is how Node.js works: you need to import the Selector class before using it.
@miherlosev 

True. You mean `require`, though right? I tried `import` and it errored (maybe because it was in a `cjs` file and not a `js` file? Although I thought node.js uses CommonJS by default anyway, so I wouldn't expect that to make a difference).

**Comment by Aleksey28** [comment]
Hi @DanKaplanSES,

It is the same. import is used in ESM and require is used in CommonJS modules, but they serve the same goal - import one module to another. In the config file, you need to use require because TestCafe doesn't support the ESM format in the config. In tests, you can use both ways.
